File: discord-bot/src/meeting/transcript.js
export function formatTimestamp(ms) {
  const total = Math.max(0, Math.floor(ms / 1000));
  const m = Math.floor(total / 60);
  const s = total % 60;
  return `${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
}

export function assembleTranscript(segments) {
  return (segments ?? [])
    .filter((seg) => (seg.text ?? '').trim().length > 0)
    .map((seg, i) => ({ ...seg, i }))
    .sort((a, b) => a.startMs - b.startMs || a.i - b.i)
    .map((seg) => `[${formatTimestamp(seg.startMs)}] ${seg.speaker}: ${seg.text.trim()}`)
    .join('\n');
}

File: services/meeting/src/pipeline/transcript.py
def format_timestamp(ms: int) -> str:
    total = max(0, ms // 1000)
    return f"{total // 60:02d}:{total % 60:02d}"


def _get(seg, key):
    return seg[key] if isinstance(seg, dict) else getattr(seg, key)


def assemble_transcript(segments) -> str:
    indexed = [(i, s) for i, s in enumerate(segments or []) if str(_get(s, "text")).strip()]
    indexed.sort(key=lambda pair: (_get(pair[1], "start_ms"), pair[0]))
    return "\n".join(
        f"[{format_timestamp(_get(s, 'start_ms'))}] {_get(s, 'speaker')}: {str(_get(s, 'text')).strip()}"
        for _, s in indexed
    )
